pkg/pubsubsource/pubsub_test.go

```diff
@@ -0,0 +1,240 @@
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	if err := pool.Retry(func() error {
+		pubsubClient, err = pubsub.NewClient(ctx, Project)
+		if err != nil {
+			log.Fatalf("Failed to create client: %v", err)
+		}
+		return nil
+	}); err != nil {
+		if resource != nil {
+			_ = pool.Purge(resource)
+		}
+		log.Fatalf("could not connect to gcloud pubsub %s", err)
+	}
+	defer pubsubClient.Close()
+	defer cancel()
+	code := m.Run()
+	if resource != nil {
+		if err := pool.Purge(resource); err != nil {
+			log.Fatalf("Couln't purge resource %s", err)
+		}
+	}
+	os.Exit(code)
+}
+
+func TestPubSubSource_Read(t *testing.T) {
+	setupCtx, cancelSetup := context.WithCancel(context.Background())
+	err := ensureTopicAndSubscription(setupCtx, pubsubClient, TopicID, subscriptionID)
+	assert.Nil(t, err)
+	cancelSetup()
+	messageCh := make(chan sourcer.Message, 20)
+	subscription := pubsubClient.Subscription(subscriptionID)
+	pubsubSource := NewPubSubSource(pubsubClient, subscription, MAX_EXTENSION_PERIOD, MAX_OUT_STANDING_MESSAGES)
+	ctx, cancel := context.WithCancel(context.Background())
+	publishChan := make(chan struct{})
+	defer cancel()
+	pubsubSource.StartReceiving(ctx)
+	go func() {
+		sendMessages(ctx, pubsubClient, TopicID, 100)
+		close(publishChan)
+
+	}()
+
+	pubsubSource.Read(ctx, mocks.ReadRequest{
+		CountValue: 5,
+		Timeout:    20 * time.Second,
+	}, messageCh)
+	assert.Equal(t, 5, len(messageCh))
+
+	// Try reading 4 more messages
+	// Since the previous batch didn't get acked, the data source shouldn't allow us to read more messages
+	// We should get 0 messages, meaning the channel only holds the previous 5 messages
+	pubsubSource.Read(ctx, mocks.ReadRequest{
+		CountValue: 4,
+		Timeout:    10 * time.Second,
+	}, messageCh)
+	assert.Equal(t, 5, len(messageCh)) // Expecting no new messages
+
+	// acknowledging the first batch
+	msg1 := <-messageCh
+	msg2 := <-messageCh
+	msg3 := <-messageCh
+	msg4 := <-messageCh
+	msg5 := <-messageCh
+	pubsubSource.Ack(ctx, mocks.TestAckRequest{
+		OffsetsValue: []sourcer.Offset{msg1.Offset(), msg2.Offset(), msg3.Offset(), msg4.Offset(), msg5.Offset()},
+	})
+
+	pubsubSource.Read(ctx, mocks.ReadRequest{
+		CountValue: 6,
+		Timeout:    10 * time.Second,
+	}, messageCh)
+	assert.Equal(t, 6, len(messageCh))
+
+	<-publishChan
+}
+
+func TestPubSubSource_Pending(t *testing.T) {
+	setupCtx, cancelSetup := context.WithCancel(context.Background())
+	err := ensureTopicAndSubscription(setupCtx, pubsubClient, TopicID, subscriptionID)
+	assert.Nil(t, err)
+	cancelSetup()
+	subscription := pubsubClient.Subscription(subscriptionID)
+	pubsubSource := NewPubSubSource(pubsubClient, subscription, MAX_EXTENSION_PERIOD, MAX_OUT_STANDING_MESSAGES)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	pubsubSource.StartReceiving(ctx)
+	sendMessages(ctx, pubsubClient, TopicID, 40)
+	val := pubsubSource.Pending(context.Background())
+	assert.Equal(t, MAX_OUT_STANDING_MESSAGES, int(val))
```

Comment:
Use a more real example. e.g. send 1000 messages, read 600 and verify pending is 400.

Reply:
That's not how it will work as the sending here is synchronous ,so lets say if I send 1000 messages and he have a buffer of lets say 600 ,so 600 message will be stored in the buffer rest 400 will be discarded (in real systems this unacknowledged message would be redelivered by the pub sub )

Reply:
Good to know that Pub/Sub is redelivering, hence we ensure no data loss.